### PR TITLE
Next Release: Version 2

### PR DESCRIPTION
**Pull request by jellydn**
## WHAT

Version 2 is coming. Please give it a try and let me know if you encounter any issues.

[IT Man - Hurl.nvim v2 Preview!](https://www.youtube.com/watch?v=B4Z3-NV_1Qo)

```lua
return {
  {
    "jellydn/hurl.nvim",
    branch = "canary",
    ft = "hurl",
    dependencies = { "MunifTanjim/nui.nvim", "nvim-treesitter/nvim-treesitter" },
    opts = {
      mode = "split",
      auto_close = false,
      debug = true,
      show_notification = false,
      formatters = {
        json = { "jq" },
        html = {
          "prettier",
          "--parser",
          "html",
        },
      },
      fixture_vars = {
        {
          name = "random_int_number",
          callback = function()
            return math.random(1, 1000)
          end,
        },
        {
          name = "random_float_number",
          callback = function()
            local result = math.random() * 10
            return string.format("%.2f", result)
          end,
        },
        {
          name = "now",
          callback = function()
            return os.date("%d/%m/%Y")
          end,
        },
      },
    },
    keys = {
      -- Run API request
      { "<leader>hA", "<cmd>HurlRunner<CR>", desc = "Run All requests" },
      { "<leader>ha", "<cmd>HurlRunnerAt<CR>", desc = "Run Api request" },
      { "<leader>he", "<cmd>HurlRunnerToEntry<CR>", desc = "Run Api request to entry" },
      { "<leader>hE", "<cmd>HurlRunnerToEnd<CR>", desc = "Run Api request from current entry to end" },
      { "<leader>hv", "<cmd>HurlVerbose<CR>", desc = "Run Api in verbose mode" },
      { "<leader>hV", "<cmd>HurlVeryVerbose<CR>", desc = "Run Api in very verbose mode" },
      { "<leader>hr", "<cmd>HurlRerun<CR>", desc = "Rerun last command" },
      -- Run Hurl request in visual mode
      { "<leader>hh", ":HurlRunner<CR>", desc = "Hurl Runner", mode = "v" },
      -- Show last response
      { "<leader>hh", "<cmd>HurlShowLastResponse<CR>", desc = "History", mode = "n" },
      -- Manage variable
      { "<leader>hg", ":HurlSetVariable", desc = "Add global variable" },
      { "<leader>hG", "<cmd>HurlManageVariable<CR>", desc = "Manage global variable" },
      -- Toggle
      { "<leader>tH", "<cmd>HurlToggleMode<CR>", desc = "Toggle Hurl Split/Popup" },
    },
  }
}
```

## WHY

- Enhancing the usability and functionality of Hurl.nvim for developers interacting with HTTP APIs within Neovim.

## HOW

- Developed new Lua commands for HurlRerun, HurlVeryVerbose, and HurlJson.
- Enabled auto-save of capture data to global variables, enhancing the persistence of session data.
- Refactored the UI for response displays, adding an option to toggle between split and popup modes.

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [x] Linter
- [x] Tests
- [ ] Review comments
- [ ] Security

----

> [!IMPORTANT]
> Enhances `hurl.nvim` with new commands, improved response handling, and UI updates, including response history management and parsing logic validation.
> 
>   - **Behavior**:
>     - New commands `HurlRerun`, `HurlVeryVerbose`, and `HurlJson` added for enhanced request handling.
>     - Auto-save of capture data to global variables enabled.
>     - UI refactored to toggle between split and popup modes.
>   - **Parsing and Execution**:
>     - `hurl_parser.lua` added for parsing Hurl command output.
>     - `hurl_runner.lua` updated to handle verbose and JSON modes.
>   - **History and State Management**:
>     - Response history management added to store last 10 responses.
>     - `get_last_response()` and `update_history()` functions added for response retrieval and update.
>   - **Testing**:
>     - New tests in `hurl_parser_spec.lua` to validate parsing logic.
>   - **Misc**:
>     - `README.md` updated with new command descriptions.
>     - `.luarc.json` updated to include `vim` in globals.
> 
> <sup>This description was created by </sup>[Ellipsis](https://www.ellipsis.dev?ref=jellydn%2Fhurl.nvim&utm_source=github&utm_medium=referral)<sup> for e069d314111e6f161a287c800d6d4e04d9df7006. It will automatically update as commits are pushed.</sup>

----

## Release Notes

- **New Features**
	- Added a new key mapping `<leader>tV` for running API requests in very verbose mode.
	- Introduced a virtual text annotation feature for HTTP entries in Hurl files.
	- Implemented response history management, allowing users to view the last 10 responses.
	- New configuration option `save_captures_as_globals` added for enhanced user settings.
	- Enhanced logging and error handling for executing Hurl commands.
	- Added a health check function to verify the presence of required executables and check Hurl version.
	- New HTTP request examples for interacting with a todo API.
	- Enhanced API request handling for retrieving dog breeds with additional data capture.

- **Improvements**
	- Streamlined popup management with improved content display and layout.
	- Refined output formatting for various data types, including markdown support.
	- Enhanced error handling for spinner notifications.

- **Bug Fixes**
	- Resolved issues related to buffer validity checks and popup closing logic.

- **Tests**
	- Added comprehensive unit tests for the `hurl_parser` module to ensure robust output parsing.

### CODE REVIEW

**Comment by jellydn** [comment]
This pull request introduces several enhancements and fixes across multiple files in the project. The changes improve the handling of responses, enhance the user interface, and update documentation and configurations. Below is a summary of the most important changes:

### Enhancements to Response Handling:
* [`lua/hurl/history.lua`](diffhunk://#diff-f455f204a00ce80a9632bf240d13ed5244575b90d4e4e06ca087254e2ac46eefL4-R53): Implemented a response history feature to store the last 10 responses and added functions to show the last response and update the history after each request.
* [`lua/hurl/lib/hurl_parser.lua`](diffhunk://#diff-0526e5c357317c3703555f7966eb93414f5827686e457c68509fe26d7dfd562cR1-R113): Added a new parser to handle Hurl output, including request details, response headers, timings, captures, and errors.

### User Interface Improvements:
* [`lua/hurl/popup.lua`](diffhunk://#diff-8f27c3f2cca529ab40bc647326bda672682fa823f6e46613e949127457cf32b5R8-R18): Refactored the popup display to separate request information and body content into different sections, and updated the layout and mappings for better navigation. [[1]](diffhunk://#diff-8f27c3f2cca529ab40bc647326bda672682fa823f6e46613e949127457cf32b5R8-R18) [[2]](diffhunk://#diff-8f27c3f2cca529ab40bc647326bda672682fa823f6e46613e949127457cf32b5L27-R30) [[3]](diffhunk://#diff-8f27c3f2cca529ab40bc647326bda672682fa823f6e46613e949127457cf32b5L46-R148) [[4]](diffhunk://#diff-8f27c3f2cca529ab40bc647326bda672682fa823f6e46613e949127457cf32b5L159-R173)
* [`lua/hurl/split.lua`](diffhunk://#diff-8662a61da78d58049b714b3f1d91e5230dc306871ab781adb4fa3af6121675c4R8-R90): Updated the split view to include request information, headers, and body content, and improved the handling of the response display. [[1]](diffhunk://#diff-8662a61da78d58049b714b3f1d91e5230dc306871ab781adb4fa3af6121675c4R8-R90) [[2]](diffhunk://#diff-8662a61da78d58049b714b3f1d91e5230dc306871ab781adb4fa3af6121675c4L91-R112)

### Configuration and Documentation Updates:
* [`.luarc.json`](diffhunk://#diff-9b02c3fd1241b15367d145b28b8742aed7d18c689ee30d8b6882fb03da98ed98L2-R3): Updated the Lua configuration to include `vim` in the global diagnostics and removed disabled diagnostics.
* [`README.md`](diffhunk://#diff-b335630551682c19a781afebcf4d07bf978fb1f8ac04c6bf87428ed5106870f5R78): Added a new keybinding for running API requests in very verbose mode and updated the verbose mode documentation with a new image. [[1]](diffhunk://#diff-b335630551682c19a781afebcf4d07bf978fb1f8ac04c6bf87428ed5106870f5R78) [[2]](diffhunk://#diff-b335630551682c19a781afebcf4d07bf978fb1f8ac04c6bf87428ed5106870f5L203-R206)

### Code Refactoring:
* [`lua/hurl/codelens.lua`](diffhunk://#diff-86c361fc2bc073331d74a4c69186135009b4738fa557677197e84af58e2d6d2aR1-R35): Added a function to insert virtual text for Hurl entries, improving the readability of HTTP verbs in the code.
* [`lua/hurl/spinner.lua`](diffhunk://#diff-f98b43a401eeaaa6c4c35feb2d904b2b58382ecef75394f9374954a0553971f6R10-R12): Enhanced the spinner functionality to show notifications when done and ensured the buffer validity before updating the spinner frames. [[1]](diffhunk://#diff-f98b43a401eeaaa6c4c35feb2d904b2b58382ecef75394f9374954a0553971f6R10-R12) [[2]](diffhunk://#diff-f98b43a401eeaaa6c4c35feb2d904b2b58382ecef75394f9374954a0553971f6R61-R69) [[3]](diffhunk://#diff-f98b43a401eeaaa6c4c35feb2d904b2b58382ecef75394f9374954a0553971f6L84-R89)

These changes collectively enhance the functionality, user experience, and maintainability of the project.

**Comment by jellydn** [comment]
Hi @cenk1cenk2 Thank you for your feedback. 

re. your question
>Is this something that can be opted out or will it be like this going forward?

I believe this is better in terms of UX so this would be my preferred approach. However, I'm open if you want to add a config for adjust this one by new PR :).

**Comment by minizilla** [comment]
I think it's a bit overkill to formatted the response as a whole markdown, I suggest make it friendly for insert them inside the code block. Something like this:

```
curl 'http://localhost:8080' \
  -d '{"foo": "this is foo", "bar": "this is bar"}'

HTTP/1.1 200 OK
Content-Type: application/json
Date: Wed, 23 Oct 2024 08:12:19 GMT
Content-Length: 283

{
  "status": "OK"
}
```

File: README.md [hunk 1]
```diff
@@ -46,7 +46,15 @@ Add the following configuration to your Neovim setup with [lazy.nvim](https://gi
...
+      "nvim-treesitter/nvim-treesitter",
+      -- Optional, for markdown rendering with render-markdown.nvim
+      {
+        'MeanderingProgrammer/render-markdown.nvim',
+        opts = {
+          file_types = { "markdown" },
+        },
+        ft = { "markdown" },
+      },
   },
```
Comment (jellydn) [inline README.md:58]:
```suggestion
  }
```

**Pull request merged**